diff.py

```diff
@@ -414,6 +429,251 @@ def get_arch(arch_str: str) -> "ArchSettings":
+        return "\n".join(
+            "".join(x.ljust(self.column_width) for x in line) for line in lines
+        )
+
+
+@dataclass
+class AnsiFormatter(Formatter):
+    column_width: int
+
+    def apply(self, line: str, f: Format) -> str:
+        if f == Format.NONE:
+            return line
+        return f"{f.ansi_code}{line}{Style.RESET_ALL}"
+
+    def table(
+        self, header: Optional[Tuple[str, ...]], lines: List[Tuple[str, ...]]
+    ) -> str:
+        if header:
+            lines = [header] + lines
+        return "\n".join("".join(self.ansi_ljust(x) for x in line) for line in lines)
+
+    def ansi_ljust(self, s: str) -> str:
+        """Like s.ljust(width), but accounting for ANSI colors."""
+        needed: int = self.column_width - ansiwrap.ansilen(s)
+        if needed > 0:
+            return s + " " * needed
+        else:
+            return s
+
+
+@dataclass
+class HtmlFormatter(Formatter):
+    def apply(self, line: str, f: Format) -> str:
+        if f == Format.NONE:
+            return line
+        class_name = f.name.lower().replace("_", "-")
+        return f"<span class='{class_name}'>{line}</span>"
+
+    def table(
+        self, header: Optional[Tuple[str, ...]], lines: List[Tuple[str, ...]]
+    ) -> str:
+        if header:
+            lines = [header] + lines
+        # TODO: Make a prettier stylesheet (maybe include it externally)
+        output = """
+<style>
+table.diff {
+    border: none;
+    font-family: Monospace;
+    white-space: pre;
+}
+table.diff .immediate {
+    color: lightblue;
+}
+table.diff .stack {
+    color: yellow;
+}
+table.diff .register {
+    color: yellow;
+}
+table.diff .delay-slot {
+    font-weight: bold;
+    color: gray;
+}
+table.diff .diff-change {
+    color: lightblue;
+}
+table.diff .diff-add {
+    color: green;
+}
+table.diff .diff-remove {
+    color: red;
+}
+table.diff .source-filename {
+    font-weight: bold;
+}
+table.diff .source-function {
+    font-weight: bold;
+    text-decoration: underline;
+}
+table.diff .source-other {
+    font-style: italic;
+}
+table.diff .color-rotation-0 {
+    color: magenta;
+}
+table.diff .color-rotation-1 {
+    color: cyan;
+}
+table.diff .color-rotation-2 {
+    color: green;
+}
+table.diff .color-rotation-3 {
+    color: red;
+}
+table.diff .color-rotation-4 {
+    color: yellow;
+}
+table.diff .color-rotation-5 {
+    color: pink;
+}
+table.diff .color-rotation-6 {
+    color: blue;
+}
+table.diff .color-rotation-7 {
+    color: lime;
+}
+table.diff .color-rotation-8 {
+    color: gray;
+}
+</style>
+"""
+        output += "<table class='diff'>\n"
+        for line in lines:
+            output += "    <tr>"
+            for cell in line:
+                output += f"<td>{cell}</td>"
+            output += "</tr>\n"
+        output += "</table>\n"
+        return output
+
+    @staticmethod
+    def escape(s: str) -> str:
+        # TODO: Check that all untrusted strings are passed through this
```

Comment:
Can we make an Html class, with "trusted input" and "text" static constructors? Or maybe it should be FormattedText, with appropriate constructors within each Formatter. (I wouldn't mind "escaping" within the AnsiFormatter by changing escape codes to unicode replacement chars, not that there's actually a case for it given it runs on trusted input.)

Reply:
I really like that pattern, but it was very difficult to incorporate into `do_diff()`. That function repeatedly re-formats the same string, which was wreaking havoc on trying to track provenance. 

After sleeping on it, I think I have two ideas for making it work -- I'll try today.

Reply:
I implemented a `Text` class that tracks a list of `(plaintext, formatting)` tuples that are preserved all the way to `OutputLine`. They aren't reified into `str`s until right before they're output.

lmk what you think of the API. (Maybe there's an argument that `Formatter.table(...)` should operate on `Text`s instead of `str`s? I'm less worried about that boundary because it's only used in 1 place.)

Reply:
That sounds like a nice API to me!

Reply:
> That sounds like a nice API to me!

Just checking - do you mean the `Text` API as-implemented, or are you saying I *should* also change `Formatter.table(...)` to work with `Text`s in this PR?

Reply:
the `Text` API as implemented. I wouldn't mind changing Formatter.table too (would get rid of the need for ansi_ljust which is nice) but as you say it's not particularly worrying